uxarray/plot/utils.py

```diff
@@ -0,0 +1,26 @@
+import holoviews as hv
+
+
+class HoloviewsBackend:
+    """Utility class to compare and set holoviews backends for
+    visualization."""
+
+    def __init__(self):
+        self.backend = None
+
+    def compare_and_set(self, backend: str):
```

Comment:
While this function makes sense to me, a property-based implementation and its usage would make more sense I think, how about something like:

```
class HoloviewsExtension:
   def __init__(self):
        self._backend = None
   @property
     def backend(self): 
         return self._backend
       
     @backend.setter 
     def backend(self, backend): 
          if backend not in ["bokeh", "matplotlib"]:
            raise ValueError(
                f"Unsupported backend. Expected one of ['bokeh', 'matplotlib'], but received {backend}"
            )

        if backend != self._backend:
            hv.extension(backend)
            self._backend = backend

hv_extension_ref = HoloviewsExtension()
```

and then usage would look like:

```
from uxarray.plot.utils import hv_extension_ref

hv_extension_ref.backend("matplotlib")
```

1. The name `hv_extension_ref` still itches; maybe just "hv_extension" or something else?
2. The words "extension" and "backend" can be swapped if desired in the above code.

Reply:
I haven't rendered the cookbook using this branch yet, but the usage examples in our documentation are only showing the backend logo when the first plot is generated or a backend is switched.

![image](https://github.com/UXARRAY/uxarray/assets/67855069/69d36180-60e7-4be1-8261-2bbfb65a1903)


The suggestion for a property-based implementation looks really clean, but I want to hear your feedback on the implementation below, which wouldn't need to use a setter.

> The name hv_extension_ref still itches; maybe just "hv_extension" or something else?

> The words "extension" and "backend" can be swapped if desired in the above code.

Yeah, I think using `backend` instead of `extension` is much more clear.

Maybe instead of directly importing the object, we could access it through the module like so, with the following naming change

Inside of uxarray.plot.utils
```Python
...
# same implementation, renamed `compare_and_set` to `assign`
backend = HoloviewsExtension()
```

```Python
from uxarray.plot.utils import plotting_backend

uxarray.plot.utils.backend.assign("matplotlib")
```

This way, we don't need to worry about importing the object and can operate directly through the module.

Reply:
> I haven't rendered the cookbook using this branch yet, but the usage examples in our documentation are only showing the backend logo when the first plot is generated or a backend is switched.
> 
> ![image](https://private-user-images.githubusercontent.com/67855069/318504894-69d36180-60e7-4be1-8261-2bbfb65a1903.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE5OTA5NjQsIm5iZiI6MTcxMTk5MDY2NCwicGF0aCI6Ii82Nzg1NTA2OS8zMTg1MDQ4OTQtNjlkMzYxODAtNjBlNy00YmUxLTgyNjEtMmJiZmI2NWExOTAzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDAxVDE2NTc0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk5Yjg1MmNhYjkzM2I1MzBhMmY1NDAwOTE5NjAyMzQzNDU2NTgzZDU1YzY4MzFkMjI0NTFhYTkxZTEwNDYwY2MmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.1CBSfSUqCx13AUpAwTfPOTscATzo65tDQ2fYCJEA72Q)

This is great! And all of the plots that are adjacently happening with the same backend are rendering properly as I've observed, right? If you have also confirmed this, maybe there is no need to look into the cookbook.

Reply:
> The suggestion for a property-based implementation looks really clean, but I want to hear your feedback on the implementation below, which wouldn't need to use a setter.
> 
> > The name hv_extension_ref still itches; maybe just "hv_extension" or something else?
> 
> > The words "extension" and "backend" can be swapped if desired in the above code.
> 
> Yeah, I think using `backend` instead of `extension` is much more clear.
> 
> Maybe instead of directly importing the object, we could access it through the module like so, with the following naming change
> 
> Inside of uxarray.plot.utils
> 
> ```python
> ...
> # same implementation, renamed `compare_and_set` to `assign`
> backend = HoloviewsExtension()
> ```
> 
> Inside of plotting methods
> 
> ```python
> from uxarray.plot.utils import plotting_backend
> 
> uxarray.plot.utils.backend.assign("matplotlib")
> ```
> 
> This way, we don't need to worry about importing the object and can operate directly through the module.

You meant:

```
from uxarray.plot.utils import plotting_backend
 
plotting_backend.assign("matplotlib")
```
right? 

Yeah, this would also work; I am fine with either way. An additional reason for me to suggest a property was that if anytime user wants to return what backend they were on (i.e. using the property getter), but this maynot be a need or can it still be handled via holoviz's native functionality?

Reply:
Whoops, I meant the following:

```Python
# import plotting utils
import uxarray.plot.utils

# directly access object through module instead of importing object
uxarray.plot.utils.backend.assign("matplotlib")
```

Reply:
Ok, this looks great!